src/clj/witan/send/send.clj

```diff
@@ -433,110 +439,126 @@
-                          (assoc :calendar-year year)))
-                    (map :total-in-send send-output) (range initial-projection-year 3000))
-               (map (apply juxt columns))
-               (concat [(map name columns)])
-               (csv/write-csv writer))))
-      (with-open [writer (io/writer (io/file "target/output-cost.csv"))]
-        (let [columns [:calendar-year :mean :std-dev :iqr :min :low-ci :q1 :median :q3 :high-ci :max]]
-          (->> (map (fn [stats year]
-                      (-> (medley/map-vals round stats)
-                          (assoc :calendar-year year)))
-                    (map :total-cost send-output) (range initial-projection-year 3000))
-               (map (apply juxt columns))
-               (concat [(map name columns)])
-               (csv/write-csv writer))))
-      (with-open [writer (io/writer (io/file "target/output-ay-group.csv"))]
-        (let [columns [:calendar-year :ay-group :mean :std-dev :iqr :min :low-ci :q1 :median :q3 :high-ci :max]]
-          (->> (mapcat (fn [output year]
-                         (map (fn [[ay-group stats]]
-                                (-> (medley/map-vals round stats)
-                                    (assoc :ay-group ay-group :calendar-year year)))
-                              (:total-in-send-by-ay-group output))) send-output (range initial-projection-year 3000))
-               (map (apply juxt columns))
-               (concat [(map name columns)])
-               (csv/write-csv writer))))
-      (run! #(ch/sankey-transitions transitions-data % valid-settings) years)
-      (ch/ribbon-plot joiner-rates-CI "Joiner" years n-colours)
-      (ch/ribbon-plot leaver-rates-CI "Leaver" years n-colours)
-      (ch/ribbon-plot mover-rates-CI "Mover" years n-colours)))
+  [{:keys [projection send-output transition-matrix valid-setting-academic-years population]} {:keys [output]}]
+  (let [transitions-data (ds/row-maps transition-matrix)
+        transform-transitions (->> transitions-data
+                                   (map #(vector
+                                          (:academic-year-2 %)
+                                          (states/state (:need-1 %) (:setting-1 %))
+                                          (states/state (:need-2 %) (:setting-2 %))))
+                                   distinct)
+        transform-projection (->> projection
+                                  keys
+                                  (map #(vec (drop 1 %)))
+                                  distinct)]
+    (if (every? (fn [transition] (transition-present? transition transform-projection)) transform-transitions)
+      (println "Not every historic transition present in projection. Consider checking valid state input"))
+    (if (= output true)
+      (let [valid-settings (assoc (->> (ds/row-maps valid-setting-academic-years)
+                                       (reduce #(assoc %1 (:setting %2) (:setting->group %2)) {}))
```

Comment:
down the line (not too long down the line) it might be good to break this up into little functions, with the appropriate let processing close to the outputs they belong to.  would also avoid the massive diff :)
possible even going further: having a function for all the output production, so that the `if` is readable
```
(if output
  (produce-all-the-output x y z)
  send-output)
```

Reply:
good call!